### PR TITLE
Make RxJava optional

### LINKED ISSUE #396: Make RxJava Optional
Requery's design is so smart that I have changed my thought to make RxJava optional.

https://github.com/requery/requery/wiki/Using-the-EntityStore-interface

### PR DESCRIPTION
**Pull request by k-kagurazaka**
Resolve #396 

Currently, RxJava is used as default.
When developers want to opt-out RxJava, write configuration as:

```java
@GenerationOption(rxJavaSupport = false)
public class OrmaGenerationOption { }
```

Sorry for this PR looks big, however the changes are pretty simple.
I extracted RxJava helper methods from some classes into new "Rx*" classes listed below:

|without Rx methods|with Rx methods|
|---|---|
|OrmaConnection|RxOrmaConnection|
|Relation|RxRelation|
|Inserter|RxInserter|
|Selector|RxSelector|
|Updater|RxUpdater|
|Deleter|RxDeleter|
|SingleAssociation|RxSingleAssociation|
|DataSetChangedTrigger|RxDatasetChangedTrigger|

Orma processor generates classes to use "Rx*" classes instead of base one when `GenerationOption(rxJavaSupport = true)`.

Most of changes are actually "code move" but some are newly written by me, so I'm going to comment into new code pieces (it's review points).

### CODE REVIEW

File: example/build.gradle [hunk 1]
```diff
@@ -82,6 +82,10 @@ dependencies {
     annotationProcessor project(':processor')
     compile project(':library')
     compile project(':encryption')
+
+    compile 'io.reactivex.rxjava2:rxjava:2.1.0'
+    compile 'io.reactivex.rxjava2:rxandroid:2.0.1'
+
```
Comment (k-kagurazaka) [inline example/build.gradle]: Now, Orma users explicitly add RxJava dependencies into own build.gradle.

File: library/src/main/java/com/github/gfx/android/orma/SingleAssociation.java [hunk 2]
```diff
@@ -111,26 +122,27 @@ public long getId() {
...
     public Model get() throws NoValueException {
-        return single.blockingGet();
+        if (value == null) {
+            synchronized (this) {
+                if (value == null) {
+                    value = factory.call();
+                }
+            }
+        }
+        return value;
```
Comment (k-kagurazaka) [inline library/src/main/java/com/github/gfx/android/orma/SingleAssociation.java]: I rewrite lazy and cached `get()` methods via double-checked technique instead `Single#blockingGet()`.

File: library/src/main/java/com/github/gfx/android/orma/event/DataSetChangedTrigger.java [hunk 3]
```diff
@@ -17,91 +17,32 @@
...
-                WeakReference<Observer<DataSetChangedEvent<?>>> observerRef = entry.getKey();
-                if (observerRef.get() != null) {
-                    observerRef.get().onNext(new DataSetChangedEvent<>(type, selector));
-                } else {
-                    observerMap.remove(observerRef);
-                }
-            }
-        }
-    }
+    class NoOp implements DataSetChangedTrigger {
```
Comment (k-kagurazaka) [inline library/src/main/java/com/github/gfx/android/orma/event/DataSetChangedTrigger.java]: Currently, experimental data changed event feature is available only when `rxJavaSupport = true`.
I think we will implement `DataSetChangedTrigger` without RxJava before removing experimental status.

**Comment by k-kagurazaka** [comment]
All my thoughts are posted.

**Comment by gfx** [comment]
👀

**Comment by gfx** [comment]
Excellent. Please wait a minute for code review because this is somewhat a large PR.

By the way, I've invited you to the organization and I'll give you a commit permission to make branches on this repository, which gives us better collaborations.

**Comment by k-kagurazaka** [comment]
Thanks. I'm glad to continue contributions to Orma.

**Comment by k-kagurazaka** [comment]
@gfx How is the progress?

**Comment by gfx** [comment]
Oh, sorry 😭 

Will merge it in this month!

**Comment by k-kagurazaka** [comment]
Rebase to solve the conflict and add following commits:

- fix for `AssociationCondition` (This is introduced in another PR)
- re-generate example

**Comment by gfx** [comment]
Cool! Thanks.

**Comment by gfx** [comment]
I have some points to fix, but anyway thanks to PR!

**Pull request merged**